### PR TITLE
Add structured dataset type checker

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
- Add structured dataset type checker
- Update schema type checker
- Make schema compatible with the structured dataset
- Allow workflow to be like
    - Upstream (structuredDatasetType) -> downstream (schemaType)
    - Upstream (schemaType) -> downstream (structuredDatasetType)

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flytekit/pull/785

## Follow-up issue
_NA_

### CODE REVIEW

**Review by EngHabu — APPROVED** [review]
Looks great! thank you
